### PR TITLE
Optimize price script.

### PR DESCRIPTION
**Pull request by passabilities**
The goal here is to decrease memory and CPU usage and increase speed.

* Separates some functions into separate files
* Queries user data from CouchDB using a Design Document View
* Finds all devices tokens for a currency hour change and batch sends notifications.
* Removes the need to store a price map and users array in memory.
* Remove RateServer class and just use axios object

Manually queuing the db was slow. Using the Design Document View added indices for _much_ faster response times and reduces the amount of information we store in memory at any given time.

### CODE REVIEW

File: src/NotificationManager.ts [hunk 1]
```diff
@@ -43,6 +43,13 @@ export class NotificationManager {
       tokens
     }
 
-    return await this.app.messaging().sendMulticast(message)
+    try {
+      const response = await this.app.messaging().sendMulticast(message)
+      console.log('Messages sent:', response)
```
Comment (passabilities) [inline src/NotificationManager.ts:48]: remove

File: src/index.ts [hunk 2]
```diff
@@ -35,92 +49,90 @@ async function run() {
...
+      // todo report error
+      console.error(error)
+    }
+  }
+
+  const thresholdList = await CurrencyThreshold.table.list()
+  for (const { id: currencyCode } of thresholdList.rows) {
+    thresholdMap[currencyCode] = await CurrencyThreshold.fetch(currencyCode)
+    await fetchPrices(currencyCode)
+  }
```
Comment (passabilities) [inline src/index.ts]: Promise.all

File: src/index.ts [hunk 3]
```diff
@@ -35,92 +49,90 @@ async function run() {
...
+          const enabled = currencyCodes[currencyCode][hours]
+          if (!enabled) continue
 
-      const userHourSettings = user.notifications.currencyCodes[currencyCode]
-      for (const [ hours, enabled ] of Object.entries(userHourSettings)) {
-        if (enabled && map[hours]) {
-          map[hours].deviceTokens.push(...deviceTokens)
-          const set = new Set(map[hours].deviceTokens)
-          map[hours].deviceTokens = Array.from(set)
+          await sendNotification(priceMap[currencyCode][hours], deviceTokens)
```
Comment (passabilities) [inline src/index.ts]: Promise.all

File: src/index.ts [hunk 4]
```diff
@@ -35,92 +49,90 @@ async function run() {
...
-    price: number
-    priceChange: number
-    timestamp: number
-    deviceTokens: Array<string>
-  }
+  [hours: string]: NotificationPriceChange
 }
 
-function sleep(ms = 5000) {
+function sleep(ms = 1000) {
```
Comment (passabilities) [inline src/index.ts]: add constant var

File: src/ratesServer.ts [hunk 5]
```diff
@@ -10,7 +10,7 @@ class RatesServer {
   }
 
   async get (path: string) {
-    const { data } = await axios(`${this.uri}/v${this.version}/${path}`)
+    const { data } = await axios(`${this.uri}/v${this.version}/${path}`, { timeout: 10000 })
```
Comment (passabilities) [inline src/ratesServer.ts]: constant var

File: src/models/base.ts [hunk 6]
```diff
@@ -50,6 +50,17 @@ export class Base implements ReturnType<typeof IModelData> {
     return item
   }
 
+  public static async fetchAll<T extends typeof Base>(this: InstanceClass<T>, ids: string[]): Promise<Array<InstanceType<T>>> {
+    const items: InstanceType<T>[] = []
+
+    for (const id of ids) {
+      const item = await this.fetch(id)
```
Comment (passabilities) [inline src/models/base.ts:57]: bulk fetch with `{ keys: ids }`

File: src/models/base.ts [hunk 7]
```diff
@@ -83,11 +94,11 @@ export class Base implements ReturnType<typeof IModelData> {
...
   }
 
-  public static async where<T extends typeof Base>(this: InstanceClass<T>, where?: Nano.MangoQuery): Promise<Array<InstanceType<T>>> {
+  public static async where<T extends typeof Base>(this: InstanceClass<T>, where?: Nano.MangoQuery, pageSize = 25, page = 0): Promise<Array<InstanceType<T>>> {
     try {
       const response = await this.table.find({
-        // NOTE: default limit to super high
-        limit: 100000000000,
+        limit: pageSize,
+        skip: page * pageSize,
```
Comment (passabilities) [inline src/models/base.ts:101]: remove

File: src/price-script/checkPriceChanges.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    const symbol = priceChange > 0 ? '+' : ''
+    const time = Number(hourChange) === 1 ? '1 hour' : `${hourChange} hours`
+
+    const title = 'Price Alert'
+    const body = `${currencyCode} is ${direction} ${symbol}${priceChange}% to $${price} in the last ${time}.`
+    const data = {}
+
+    console.log('=================')
+    console.log(`Sending ${deviceTokens.length} notifications for ${currencyCode}.`)
+    console.log('=================')
```
Comment (passabilities) [inline src/price-script/checkPriceChanges.ts:42]: remove

File: src/price-script/checkPriceChanges.ts [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+      if (usersDevices.length === 0) continue
+
+      const deviceIds: string[] = []
+      for (const { value: userDevices } of usersDevices) {
+        for (const deviceId in userDevices) {
+          deviceIds.push(deviceId)
+        }
+      }
+      const tokenGenerator = deviceTokenGenerator(deviceIds)
+      while(true) {
```
Comment (passabilities) [inline src/price-script/checkPriceChanges.ts:68]: use while(done)

File: src/price-script/checkPriceChanges.ts [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        if (done) break
+      }
+    }
+  }
+}
+
+async function* deviceTokenGenerator(deviceIds: string[]): AsyncGenerator<string[], string[]> {
+  let tokens: string[] = []
+  let bookmark: string
+  while (true) {
```
Comment (passabilities) [inline src/price-script/checkPriceChanges.ts:85]: use the if logic below

File: src/price-script/fetchThresholdPrices.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+const HOURS_PERCENT_MAP = {
+  1: 3,
+  24: 10
+}
+
+interface IThresholdPricesResponse {
+  [hours: string]: NotificationPriceChange
+}
+
+function sleep(ms = 1000) {
```
Comment (passabilities) [inline src/price-script/fetchThresholdPrices.ts:14]: move to constant

File: src/price-script/fetchThresholdPrices.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    return response
+  }
+
+  for (const hours in currencyThreshold.thresholds) {
+    const hoursAgo = Date.now() - (Number(hours) * 60 * 60 * 1000)
+    let threshold = currencyThreshold.thresholds[hours]
+    let before
+    if (threshold.lastUpdated === 0)
+      before = hoursAgo
+    else if (hoursAgo > threshold.lastUpdated)
```
Comment (passabilities) [inline src/price-script/fetchThresholdPrices.ts:37]: combine

File: src/price-script/fetchThresholdPrices.ts [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+          console.error(`Could not update ${currencyCode} threshold data.`)
+          console.error(err)
+        })
+
+      // Set decimal place to 2 significant digits
+      let numSplit = price.toString().split('.')
+      let sigIndex = numSplit[1].search(/[1-9]/)
+      numSplit[1] = numSplit[1].substring(0, sigIndex + 2)
+      const displayPrice = Number(numSplit.join('.'))
+      priceData.price = displayPrice
```
Comment (passabilities) [inline src/price-script/fetchThresholdPrices.ts:76]: move to function and unit test

**Review by thehobbit85 — CHANGES_REQUESTED** [review]
The ones that Noah added

**Review by thehobbit85 — APPROVED** [review]

**Pull request merged**